File: Editor/src/main/java/org/turbomedia/turboedit/editor/integration/DiscordIntegration.java
package org.turbomedia.turboedit.editor.integration;

import de.jcm.discordgamesdk.Core;
import de.jcm.discordgamesdk.CreateParams;
import de.jcm.discordgamesdk.LogLevel;
import de.jcm.discordgamesdk.Result;
import de.jcm.discordgamesdk.activity.Activity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.turbomedia.turboedit.editor.events.EventSystem;
import org.turbomedia.turboedit.editor.events.EventType;

import java.time.Instant;

import static org.turbomedia.turboedit.editor.ProjectManager.CURRENT_PROJECT;

public class DiscordIntegration extends Thread {
    private final static long APPLICATION_ID = 1428707164876312607L;
    private final static Logger logger = LoggerFactory.getLogger(DiscordIntegration.class);

    public DiscordIntegration() {
        super("DiscordIntegration");
    }

    @Override
    public void run() {
        super.run();

        try (var params = new CreateParams()) {
            params.setClientID(APPLICATION_ID);
            params.setFlags(CreateParams.getDefaultFlags());

            try (var core = new Core(params)) {
                core.setLogHook(LogLevel.VERBOSE, (level, message) -> {
                    switch (level) {
                        case INFO -> logger.info(message);
                        case WARN -> logger.warn(message);
                        case ERROR -> logger.error(message);
                        case DEBUG -> logger.debug(message);
                    }
                });

                if (!core.isDiscordRunning()) return;

                try (var activity = new Activity()) {
                    activity.timestamps().setStart(Instant.now());

                    EventSystem.RegisterListener(EventType.DISCORD_UPDATE_ACTIVITY, (dat) -> {
                        activity.setDetails(CURRENT_PROJECT.name());
                        activity.setState("Editing");
                        activity.assets().setLargeImage("app_icon");
                        activity.assets().setLargeText("TurboEdit");

                        core.activityManager().updateActivity(activity, (result) -> {
                            if (result != Result.OK) {
                                logger.error("Couldn't create discord activity");
                            }
                        });
                    });

                    EventSystem.CallEvent(EventType.DISCORD_UPDATE_ACTIVITY);
                }

                while (true) {
                    core.runCallbacks();

                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }
}

File: Editor/src/main/kotlin/turboedit/editor/integration/DiscordIntegration.kt
package turboedit.editor.integration

import de.jcm.discordgamesdk.Core
import de.jcm.discordgamesdk.CreateParams
import de.jcm.discordgamesdk.LogLevel
import de.jcm.discordgamesdk.Result
import de.jcm.discordgamesdk.activity.Activity
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import turboedit.editor.events.EventSystem
import turboedit.editor.events.EventSystem.EventAction
import turboedit.editor.events.EventType
import turboedit.editor.ProjectManager.CURRENT_PROJECT
import java.time.Instant
import java.util.function.BiConsumer
import java.util.function.Consumer

class DiscordIntegration : Thread("DiscordIntegration") {
    override fun run() {
        super.run()

        CreateParams().use { params ->
            params.setClientID(APPLICATION_ID)
            params.setFlags(CreateParams.getDefaultFlags())

            Core(params).use { core ->
                core.setLogHook(LogLevel.VERBOSE, BiConsumer { level: LogLevel?, message: String? ->
                    when (level) {
                        LogLevel.INFO -> logger.info(message)
                        LogLevel.WARN -> logger.warn(message)
                        LogLevel.ERROR -> logger.error(message)
                        LogLevel.DEBUG -> logger.debug(message)
                        else -> {}
                    }
                })

                if (!core.isDiscordRunning) return

                Activity().use { activity ->
                    activity.timestamps().start = Instant.now()

                    EventSystem.registerListener(EventType.DISCORD_UPDATE_ACTIVITY, EventAction { dat: Any? ->
                        activity.details = CURRENT_PROJECT!!.name
                        activity.state = "Editing"
                        activity.assets().largeImage = "app_icon"
                        activity.assets().largeText = "TurboEdit"

                        core.activityManager().updateActivity(activity, Consumer { result: Result? ->
                            if (result != Result.OK) {
                                logger.error("Couldn't create discord activity")
                            }
                        })
                    })

                    EventSystem.callEvent(EventType.DISCORD_UPDATE_ACTIVITY)
                }
                while (true) {
                    core.runCallbacks()

                    try {
                        sleep(1000)
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                }
            }
        }
    }

    companion object {
        private const val APPLICATION_ID = 1428707164876312607L
        private val logger: Logger = LoggerFactory.getLogger(DiscordIntegration::class.java)
    }
}
